File: C++/문제_16398번.cpp
//문제 16398번 행성 연결 -> 최소 스패닝 트리, 크루스칼
#include <vector>
vector<pair<int,pair<int,int>>>v;
int parent[1001];
int find(int node){
    if(node == parent[node])
        return node;
    return parent[node] = find(parent[node]);
}
void merge(int node1, int node2){
    node1 = find(node1);
    node2 = find(node2);
    if(node1 != node2)
        parent[node2] = node1;
}
int main(){
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    cout.tie(NULL);
    int n;
    long long result = 0;
    cin >> n;
    for(int i = 1;i<=n;i++){
        parent[i] = i;
    }
    for(int i = 1;i<=n;i++){
        for(int j = 1;j<=n;j++){
            int cost;
            cin >> cost;
            if(i<j)
                v.push_back({cost,{i, j}});
        }
    }
    sort(v.begin(), v.end());
    for(int i = 0;i<v.size();i++){
        int gx = v[i].second.first;
        int gy = v[i].second.second;
        int cost = v[i].first;
        if(find(gx) != find(gy)){
            merge(gx, gy);
            result+=cost;
        }
    }
    cout << result;
}

File: JavaScript/문제_16398번.js
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

let n = +input[0];
let arr = [];
let parent = new Array(n + 1);
const find = (node) => {
  if (parent[node] == node)
    return node;
  return parent[node] = find(parent[node]);
}
const merge = (node1, node2) => {
  node1 = find(node1);
  node2 = find(node2);
  if (node1 != node2)
    parent[node2] = node1;
}

for (let i = 1; i <= n; i++){
  input[i] = input[i].split(' ').map(o => +o);
  arr.push(input[i]);
  parent[i] = i;
}
let pq = [];//간선, from, to
for (let i = 0; i < n; i++){
  for (let j = i + 1; j < n; j++){
    pq.push({ cost: arr[i][j], from: i, to: j });
  }
}
pq.sort((o1, o2) => {
  return o1.cost - o2.cost;
});
let result = 0;
for (let i = 0; i < pq.length; i++){
  let from = pq[i].from;
  let to = pq[i].to;
  let weight = pq[i].cost;
  if (find(from) != find(to)) {
    result += weight;
    merge(from, to);
  }
}
console.log(result);